### ISSUE TITLE
Grappler errors on LSTM jacobian when `experimental_use_pfor=False`

### ISSUE THREAD

**Issue by hartikainen** [tensorflow/tensorflow#41129]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): `Yes`
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): `Ubuntu 16.04`
- TensorFlow installed from (source or binary): `Binary`
- TensorFlow version (use command below): `v1.12.1-35711-g8202ae9d5c 2.4.0-dev20200705`
- Python version: `3.7`
- CUDA/cuDNN version:
```
$ conda list | grep cud
cudatoolkit               10.1.243             h6bb024c_0
cudnn                     7.6.5                cuda10.1_0
```
- GPU model and memory: `Nvidia GeForce GTX 1080 Ti`

**Describe the current behavior**
When running with gpu, grappler errors when computing `GradientTape.jacobian(..., experimental_use_pfor=False)` through LSTM. The code below runs, but outputs error messages to stdout. In some cases, similar code seems to be slower on gpu than cpu, even though the `GradientTape.gradient` is much faster on gpu than cpu.

**Describe the expected behavior**
Code runs without grappler errors.

**Standalone code to reproduce the issue**
```python
import tensorflow as tf


batch_size, sequence_length = 3, 5

x_input = tf.keras.layers.Input(
    shape=(sequence_length, 1),
    name='input',
    dtype=tf.float32)

mask_input = tf.keras.layers.Input(
    shape=(sequence_length, ),
    name='mask',
    dtype=tf.bool)


out = tf.keras.layers.LSTM(
    units=256,
    return_sequences=True,
    return_state=False,
)(x_input, mask=mask_input)
hidden_layer_sizes = (256,  256)
for hidden_layer_size in hidden_layer_sizes:
    out = tf.keras.layers.Dense(hidden_layer_size, activation='relu')(out)
out = tf.keras.layers.Dense(1, activation='linear')(out)
model = tf.keras.Model((x_input, mask_input), out)

x = tf.random.uniform(
    (batch_size, sequence_length, x_input.shape[-1]),
    dtype=x_input.dtype)

mask = tf.sequence_mask(
    tf.random.uniform(
        (batch_size, ), minval=0, maxval=sequence_length, dtype=tf.int32),
    maxlen=sequence_length,
)[..., ::-1]


@tf.function(experimental_relax_shapes=True)
def compute_jacobian(x):
    y_true = tf.zeros(batch_size)
    with tf.GradientTape() as tape:
        y = model((x, mask))
        y = tf.reduce_sum(y, axis=1)
        loss = tf.losses.MSE(y_pred=y, y_true=y_true)

    jacobian = tape.jacobian(
        loss,
        model.trainable_variables,
        parallel_iterations=10,
        experimental_use_pfor=False)

    return jacobian

_ = compute_jacobian(x)
```

**Other info / logs**
```
$ export CUDA_VISIBLE_DEVICES=1 && python ./test_lstm_jacobian.py
...
2020-07-06 18:19:52.907774: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] function_optimizer failed: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_1/lstm/PartitionedCall:5 incompatible with expected int32.
2020-07-06 18:19:52.954497: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] layout failed: Out of range: src_output = 26, but num_outputs is only 26
2020-07-06 18:19:52.999724: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] function_optimizer failed: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_1/lstm/PartitionedCall:5 incompatible with expected int32.
2020-07-06 18:19:53.074385: W tensorflow/core/common_runtime/process_function_library_runtime.cc:773] Ignoring multi-device function optimization failure: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_1/lstm/PartitionedCall:5 incompatible with expected int32.
...
```

**Comment by ravikyram** [comment]
@hartikainen 

I have tried in colab with TF nightly versions and i am not seeing any error.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/3722c2bef5b3c2dd4158e8dfe0e1ce5e/untitled.ipynb).Please, help me to reproduce the issue so it helps me in debugging.Thanks!

**Comment by hartikainen** [comment]
@ravikyram, thanks for looking into this! How do I check the cuda drivers that are being used by tensorflow in the colab above? I'm using `cudatoolkit==10.1.243` and `cudnn==7.6.5`. Is it possible that these are different? Also, could the different GPU (`Tesla T4` on colab vs. `GTX 1080 Ti` on my end) make difference here?

**Comment by hartikainen** [comment]
@ravikyram, looking into this again, I do actually see the same errors in your colab. If you click `Runtime -> View runtime logs`, you'll see:
> Jul 7, 2020, 3:12:20 PM | WARNING | 2020-07-07 12:12:20.047368: W tensorflow/core/common_runtime/process_function_library_runtime.cc:773] Ignoring multi-device function optimization failure: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_5/lstm_2/PartitionedCall:5 incompatible with expected int32.
-- | -- | --
Jul 7, 2020, 3:12:20 PM | WARNING | 2020-07-07 12:12:20.003220: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] function_optimizer failed: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_5/lstm_2/PartitionedCall:5 incompatible with expected int32.
Jul 7, 2020, 3:12:19 PM | WARNING | 2020-07-07 12:12:19.976969: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] layout failed: Out of range: src_output = 26, but num_outputs is only 26
Jul 7, 2020, 3:12:19 PM | WARNING | 2020-07-07 12:12:19.953217: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:581] function_optimizer failed: Invalid argument: Input 0 of node while/enter/_25 was passed bool from functional_5/lstm_2/PartitionedCall:5 incompatible with expected int32.
Jul 7, 2020, 3:12:07 PM | INFO | Adapting to protocol v5.1 for kernel fe733757-dd0f-411c-a530-edf26d581989

**Comment by abattery** [comment]
Sorry. Grappler is not a TFLite component. Please find owner(s) of the grappler.

**Comment by saikumarchalla** [comment]
Issue still exists in TF 2.5 as well  and I can see as warnings in  logs. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/d4362e264996d6c54a5bbb2f6b87938a/untitled39.ipynb#scrollTo=bgS959wM5Hs9&uniqifier=3).Thanks!

**Comment by sushreebarsa** [comment]
@hartikainen I tried to run the code on colab using TF v2.7.0 and didn't  face the error reported.Could you please find the attached gist [here](https://colab.research.google.com/gist/sushreebarsa/26770d316d6ae751a4cfa4dd4f9998cb/41129.ipynb) and let us know if it helps?Thanks!

**Comment by hartikainen** [comment]
Hey @sushreebarsa. I briefly tested this on my end as well and everything seems to be good. Thanks a lot for the fix 🙏

**Comment by hartikainen** [comment]
Apologies, looking this a bit more carefully, I'm still seeing grappler errors from the above code. It's a bit different than the original error but might be related:
```
python ./tests/test_tf_lstm_grappler.py
2022-01-24 10:03:14.491930: I tensorflow/core/platform/cpu_feature_guard.cc:151] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2022-01-24 10:03:26.419096: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 954 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:1a:00.0, compute capability: 7.5
2022-01-24 10:03:26.421037: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:1 with 954 MB memory:  -> device: 1, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:1b:00.0, compute capability: 7.5
2022-01-24 10:03:26.423187: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:2 with 9643 MB memory:  -> device: 2, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:60:00.0, compute capability: 7.5
2022-01-24 10:03:26.424719: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:3 with 954 MB memory:  -> device: 3, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:61:00.0, compute capability: 7.5
2022-01-24 10:03:26.426683: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:4 with 6934 MB memory:  -> device: 4, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:b1:00.0, compute capability: 7.5
2022-01-24 10:03:26.428275: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:5 with 954 MB memory:  -> device: 5, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:b2:00.0, compute capability: 7.5
2022-01-24 10:03:26.430053: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:6 with 2177 MB memory:  -> device: 6, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:da:00.0, compute capability: 7.5
2022-01-24 10:03:26.431923: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1525] Created device /job:localhost/replica:0/task:0/device:GPU:7 with 9643 MB memory:  -> device: 7, name: NVIDIA GeForce RTX 2080 Ti, pci bus id: 0000:db:00.0, compute capability: 7.5
2022-01-24 10:03:28.250137: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] function_optimizer failed: INVALID_ARGUMENT: Input 0 of node while/enter/_25 was passed bool from model/lstm/PartitionedCall:5 incompatible with expected int32.
2022-01-24 10:03:28.283472: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] layout failed: OUT_OF_RANGE: src_output = 27, but num_outputs is only 27
2022-01-24 10:03:28.318619: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] tfg_optimizer{} failed: INVALID_ARGUMENT: Input 0 of node while/enter/_25 was passed bool from model/lstm/PartitionedCall:5 incompatible with expected int32.
        when importing GraphDef to MLIR module in GrapplerHook
2022-01-24 10:03:28.329970: E tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] function_optimizer failed: INVALID_ARGUMENT: Input 0 of node while/enter/_25 was passed bool from model/lstm/PartitionedCall:5 incompatible with expected int32.
2022-01-24 10:03:28.383351: W tensorflow/core/common_runtime/process_function_library_runtime.cc:866] Ignoring multi-device function optimization failure: INVALID_ARGUMENT: Input 0 of node while/enter/_25 was passed bool from model/lstm/PartitionedCall:5 incompatible with expected int32.
```

**Comment by sushreebarsa** [comment]
@hartikainen Could you please have a look at the [gist](https://colab.research.google.com/gist/sushreebarsa/26770d316d6ae751a4cfa4dd4f9998cb/41129.ipynb#scrollTo=YXIUTGpZJWVB) on GPU runtime as well.Thanks!

**Comment by hartikainen** [comment]
Hey @sushreebarsa. Just tested on the GPU colab and it still fails. I think you might need to refresh the runtime logs after you run the code.

See the error log from the [GPU Colab](https://colab.research.google.com/gist/sushreebarsa/26770d316d6ae751a4cfa4dd4f9998cb/41129.ipynb#scrollTo=YXIUTGpZJWVB) below:
<details>
  <summary>(Collapsed)</summary>
```
> Jan 25, 2022, 11:29:38 AM | WARNING | 2022-01-25  11:29:38.658496: W  tensorflow/core/common_runtime/process_function_library_runtime.cc:866]  Ignoring multi-device function optimization failure: INVALID_ARGUMENT:  Input 0 of node while/enter/_25 was passed bool from  model_1/lstm_1/PartitionedCall:5 incompatible with expected int32.
-- | -- | --
Jan 25, 2022, 11:29:38 AM | WARNING | 2022-01-25  11:29:38.610777: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  function_optimizer failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model_1/lstm_1/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:29:38 AM | WARNING | when importing GraphDef to MLIR module in GrapplerHook
Jan 25, 2022, 11:29:38 AM | WARNING | 2022-01-25  11:29:38.600701: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  tfg_optimizer{} failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model_1/lstm_1/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:29:38 AM | WARNING | 2022-01-25  11:29:38.574004: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] layout  failed: OUT_OF_RANGE: src_output = 27, but num_outputs is only 27
Jan 25, 2022, 11:29:38 AM | WARNING | 2022-01-25  11:29:38.550275: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  function_optimizer failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model_1/lstm_1/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:28:56 AM | WARNING | 2022-01-25  11:28:56.007342: W  tensorflow/core/common_runtime/process_function_library_runtime.cc:866]  Ignoring multi-device function optimization failure: INVALID_ARGUMENT:  Input 0 of node while/enter/_25 was passed bool from  model/lstm/PartitionedCall:5 incompatible with expected int32.
Jan 25, 2022, 11:28:55 AM | WARNING | 2022-01-25  11:28:55.807796: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  function_optimizer failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model/lstm/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:28:55 AM | WARNING | when importing GraphDef to MLIR module in GrapplerHook
Jan 25, 2022, 11:28:55 AM | WARNING | 2022-01-25  11:28:55.770614: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  tfg_optimizer{} failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model/lstm/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:28:55 AM | WARNING | 2022-01-25  11:28:55.657697: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812] layout  failed: OUT_OF_RANGE: src_output = 27, but num_outputs is only 27
Jan 25, 2022, 11:28:55 AM | WARNING | 2022-01-25  11:28:55.601893: E  tensorflow/core/grappler/optimizers/meta_optimizer.cc:812]  function_optimizer failed: INVALID_ARGUMENT: Input 0 of node  while/enter/_25 was passed bool from model/lstm/PartitionedCall:5  incompatible with expected int32.
Jan 25, 2022, 11:28:49 AM | WARNING | 2022-01-25  11:28:49.566799: W  tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:39] Overriding  allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment  variable is set. Original config value was 0.
Jan 25, 2022, 11:28:39 AM | INFO | Adapting to protocol v5.1 for kernel 66de4e16-db78-47f5-8419-106c9fdb1741
```
</details>

**Comment by gadagashwini** [comment]
@hartikainen, I tried with tf-nightly version on GPU runtime, but i didn’t see any error log.
Kindly take a look at [gist](https://colab.research.google.com/drive/1QIAqMtDoNmOOA2jfUqB4-1_hjq7u8-dd?usp=sharing) and confirm issue. Thank you!

**Comment by hartikainen** [comment]
Hey @gadagashwini. Thanks for checking this. I ran the gist and can't see the error either, so I'll close this one. Thanks for the fix!
